1.0.0-rc.1

splatteredbits opened the pull request:
Finally. Feature complete.

splatteredbits committed:
Adding function Set-CIisWebsiteLogFile for configuring a website's log file settings.

splatteredbits committed:
Adding function Set-CIisAppPoolRecyclingPeriodicRestart for configuring an IIS application pool's periodic restart settings.

splatteredbits committed:
It's nice to have a default applicationHost.config file to look at.

splatteredbits committed:
Renaming Get-CIisAppPoolRecyclingPeriodicRestart to Get-CIisAppPoolPeriodicRestart because shorter is better.

splatteredbits committed:
Creating function Set-CIisAppPool for configuring base application pool settings (i.e. attributes on an application pool's add element in applicationHost.config).

splatteredbits committed:
Publish only prerelease versions for now.

splatteredbits committed:
Fixed: prerelease missing from .psd1 in .zip file.

decoyjoe commented on Carbon.IIS/Functions/Install-CIisAppPool.ps1:
```diff
@@ -3,94 +3,73 @@ function Install-CIisAppPool
-    versions `v1.0`, `v1.1`, `v2.0`, and `v4.0` are supported. Use an empty string if you're running .NET Core or to set
-    the .NET framework version to `No Managed Code`.
+    If you pass just a name, the function creates a 64-bit application pool that runs an integrated pipeline using
+    the .NET 4.0 managed runtime. To use a 32-bit app pool, use the `Enable32BitApps` . To use a classic pipeline,
```
Typo: `Enable32BitApps` should be `Enable32BitAppOnWin64`

decoyjoe commented on Carbon.IIS/Functions/Install-CIisAppPool.ps1:
```diff
@@ -3,94 +3,73 @@ function Install-CIisAppPool
 
-    To run an application pool using the classic pipeline mode, set the `ClassicPipelineMode` switch.
+    To configure the application pool's process model (i.e. the application pool's account/identity, idle timeout, etc.,
+    use the `Set-CIisAppPoolProcessModel`)
```
You're missing the word "function" at the end of the sentence. Plus, the closing paren `)` looks to be in the wrong spot.

decoyjoe commented on Carbon.IIS/Functions/Install-CIisAppPool.ps1:
```diff
@@ -3,94 +3,73 @@ function Install-CIisAppPool
-
-    Creates or sets the Cyberdyne app pool to run as NetworkService, in 32-bit mode (i.e. 32-bit applications are
-    enabled), using the classic IIS request pipeline.
+    Install-CIisAppPool -Name Cyberdyne -Enable32BitApps -ClassicPipelineMode -ManagedRuntime 'v2.0'
```
* `Enable32BitApps` should be `Enable32BitAppOnWin64`
* `ClassicPipelineMode` is not a valid parameter
* `ManagedRuntime` is not a valid parameter

decoyjoe commented on Carbon.IIS/Functions/Install-CIisAppPool.ps1:
```diff
@@ -3,94 +3,73 @@ function Install-CIisAppPool
-    the .NET framework version to `No Managed Code`.
+    If you pass just a name, the function creates a 64-bit application pool that runs an integrated pipeline using
+    the .NET 4.0 managed runtime. To use a 32-bit app pool, use the `Enable32BitApps` . To use a classic pipeline,
+    use the `ClassicPipelineMode` switch. To use a different version of .NET, use the `ManagedRuntimeVersion`
```
`ClassicPipelineMode` is not a valid parameter.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1:
```diff
@@ -0,0 +1,135 @@
+    Use the `Schedule` parameter to add times to the periodic restart configuration for time each day IIS should recycle
+    the application pool.
+
+    If any parameters are not passed, those settings will be reset to their default values.
```
> If any parameters are not passed, those settings will be reset to their default values.

Hmm, what's the reasoning for resetting omitted parameters to default settings rather than simply not configuring them? This is the opposite of what I would expect to happen if I omitted parameters.

splatteredbits replied:
The idea was that after calling most of the new functions, you'd *always* have the exact same configuration, even if someone came and modified something by hand.

Maybe add a `-All` switch that enables the current behavior and change the default behavior to only operate on passed parameters?

decoyjoe replied:
Yeah I would be more comfortable with an extra switch that enables that behavior like `-All` or `-Declarative` or something.

Normally if I'm calling `Set-CIisAppPoolPeriodicRestart -AppPoolName 'Snafu' -Memory 1000000` I would not want the `privateMemory` or `requests` settings to automatically change, especially if I did not know what they were set to beforehand and couldn't easily restore them back to their original values.

splatteredbits replied:
OK. I'll make this change in a separate PR.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisWebsiteLogFile.ps1:
```diff
@@ -0,0 +1,92 @@
+        [Parameter(Mandatory, ParameterSetName='SetInstance', Position=0)]
+        [String] $SiteName,
+
+        # If true, the function configures IIS' application pool defaults instead of
```
Incomplete sentence.

splatteredbits committed:
Fixing flaky tests.

* Install-CIisApplication tests are failing because errors from failed requests to get website content are sticking
around.
* Fixed a typo from the repo's template.
* Install-CIisWebsite tests failing because the websites under test weren't explicitly given app pools, so requests to
get content always failed.
* Created a `Get-Port` test function to avoid port number collision across tests. This didn't end up fixing any tests,
but I think its still good to not have tests use the same HTTP ports, so left it in.
* Fixed flaky Set-CIisHttpRedirect tests. They were failing because the test website wasn't assigned an app pool *and*
requests to get website content were flaky.

splatteredbits committed:
Moving enums into their own .psm1 file so tests can "using module" them.

splatteredbits committed:
Need to be able to set the IIS application pool and website defaults settings.

The objects for an application pool and the application pool defaults are the same. So are the objects for a website
and the website defaults. If we make code that gets application pools and websites also support getting defaults, any
set function we create now and in the future, don't really need to be aware if they are modifying defaults or an actual
applcation pool or website.

* Updated Get-CIisWebsite to return the site defaults settings, using the
`Microsoft.Web.Administration.ServerManager.SiteDefaults` property.
* Updated Get-CIisAppPool to return the ApplicationPoolDefaults settings, using the
`Microsoft.Web.Administration.ServerManager.ApplicationPoolDefaults` property.
* Noticed that using [switch] for boolean settings means they're always set to false. Use [bool] instead, that way
their value only changes if someone explicitly passes a value.
* Updated Set-CIisWebsiteLogFile to configure website defaults settings.
* Updated Set-CIisAppPoolCpu and Set-CIisAppPoolRecyclingPeriodicRestart to configure application pool defaults
settings.

splatteredbits committed:
Creating function Set-CIisAppPoolProcessModel for configuring an application pool or the application pool defaults
process model settings.

* Created a new template and script for auto-generating Set functions that modify sites or application pools.
* Removed process model parameters from Install-CIisAppPool in favor of using Set-CIisAppPoolProcessModel.

splatteredbits committed:
Creating function Set-CIisWebsite for configuring website settings.

These are the settings on the <site> element in the applicationHost.config file, which are currently only
`serverStartMode` and `Id`. The `Id` parameter can only be set and changed. It can never be unset/removed.

Updated `Install-CIisWebsite` to match the parameters of `Set-CIisWebsite` so users don't need to call two functions to
install and configure a new website, and to make it easier to call `Set-CIisWebsite` from `Install-CIisWebsite`.
